static/src/javascripts/projects/commercial/types.d.ts

```diff
@@ -24,3 +24,47 @@ type GuAdSize = {
 	height: number;
 	toString: (_: void) => string;
 };
+
+type RuleSpacing = {
+	minAbove: number;
+	minBelow: number;
+};
+
+type SpacefinderItem = {
+	top: number;
+	bottom: number;
+	element: HTMLElement;
+};
+
+type SpacefinderRules = {
+	bodySelector: string;
+	body?: Node;
+	slotSelector: string;
+	// minimum from slot to top of page
+	absoluteMinAbove?: number;
+	// minimum from para to top of article
+	minAbove: number;
+	// minimum from (top of) para to bottom of article
+	minBelow: number;
+	// vertical px to clear the content meta element (byline etc) by. 0 to ignore
+	// used for carrot ads
+	clearContentMeta?: number;
+	// custom rules using selectors.
+	selectors: Record<string, RuleSpacing>;
+	// will run each slot through this fn to check if it must be counted in
+	filter?: (x: SpacefinderItem) => boolean;
+	// will remove slots before this one
+	startAt?: HTMLElement;
+	// will remove slots from this one on
+	stopAt?: HTMLElement;
+	// will reverse the order of slots (this is useful for lazy loaded content)
+	fromBottom?: boolean;
+};
+
+type SpacefinderWriter = (paras: HTMLElement[]) => void;
+
+type SpacefinderOptions = {
+	waitForLinks?: boolean;
+	waitForImages?: boolean;
+	waitForInteractives?: boolean;
+};
```

Comment:
This is probably another conversation but I am preferring to have types as close to the modules they're used by and exported / imported rather than as global types.

I feel you get better discoverability but it is totally subjective so maybe a topic for dev time.

Reply:
These types could be colocated with Spacefinder once that's converted, their [equivalents in Flow](https://github.com/guardian/frontend/blob/main/static/src/javascripts.flow.archive/projects/common/modules/spacefinder.js#L23-L84) were.

Reply:
Both good points! 
For the moment maybe these could live in `space-filler.ts` until spacefinder is converted? Also open to other suggestions.